great_expectations/types/__init__.py

```diff
@@ -66,55 +84,176 @@ def __len__(self):
                 new_dict[key] = [temp_element for temp_element in value]
                 for i, element in enumerate(value):
                     if isinstance(element, DictDot):
-                        new_dict[key][i] = element.to_dict()
-                    elif isinstance(element, Enum):
+                        new_dict[key][i] = element.to_raw_dict()
+
+                    if isinstance(element, Enum):
                         new_dict[key][i] = element.value
 
             # Note: conversion will not work automatically if there are additional layers in between.
 
         return new_dict
 
+    def to_dict(self) -> dict:
+        new_dict = {
+            key: self[key]
+            for key in self.property_names(
+                include_keys=self.include_field_names,
+                exclude_keys=self.exclude_field_names,
+            )
+        }
+        for key, value in new_dict.items():
+            if isinstance(value, DictDot):
+                new_dict[key] = value.to_dict()
+
+            if isinstance(value, Enum):
+                new_dict[key] = value.value
+
+            if isinstance(value, list) or isinstance(value, tuple):
+                new_dict[key] = [temp_element for temp_element in value]
+                for i, element in enumerate(value):
+                    if isinstance(element, DictDot):
+                        new_dict[key][i] = element.to_dict()
+
+                    if isinstance(element, Enum):
+                        new_dict[key][i] = element.value
+
+        return new_dict
+
+    def property_names(
+        self,
+        include_keys: Optional[Set[str]] = None,
+        exclude_keys: Optional[Set[str]] = None,
+    ) -> Set[str]:
+        """
+        Assuming that -- by convention -- names of private properties of an object are prefixed by "_" (a single
+        underscore character), return these property names as public property names.  To support this convention, the
+        extending classes must implement property accessors, corresponding to the property names, return by this method.
+
+        :param include_keys: inclusion list ("include only these properties, while excluding all the rest")
+        :param exclude_keys: exclusion list ("exclude only these properties, while include all the rest")
+        :return: property names, subject to inclusion/exclusion filtering
+        """
+        if include_keys is None:
+            include_keys = set()
+
+        if exclude_keys is None:
+            exclude_keys = set()
+
+        if include_keys & exclude_keys:
+            raise ValueError(
+                "Common keys between sets of include_keys and exclude_keys filtering directives are illegal."
+            )
+
+        key: str
+
+        property_names: Set[str] = set(
+            filter(
+                lambda name: len(name) > 1,
+                [key[1:] for key in self.keys() if key[0] == "_"],
+            )
+        )
```

Comment:
Could we add a comment to clarify this is meant to filter private vars?

Love the functional style here ❤️

Reply:
@cdkini I will add the comment (in the next commit).  Thanks!

Reply:
Comment added.